src/write.c

```diff
@@ -2400,129 +2358,218 @@ static avifResult avifEncoderWriteMinimizedImageBox(avifEncoder * encoder, avifR
+    flags |= fullRange ? 0x000080 : 0;                // full_range
+    flags |= pixelFormat << 8;                        // pixel_format (4 bits)
+    flags |= chromaSubsampling << 12;                 // chroma_subsampling (2 bits)
+    flags |= isHorizontallyCentered ? 0x004000 : 0;   // is_horizontally_centered
+    flags |= isVerticallyCentered ? 0x008000 : 0;     // is_vertically_centered
+    flags |= orientation << 16;                       // orientation (3 bits)
+    flags |= hasExplicitCodecTypes ? 0x080000 : 0;    // has_explicit_codec_types
+    flags |= dimensionSize == 7 ? 0x100000 : 0;       // dimension_size
+    flags |= codecConfigSizeSize == 3 ? 0x200000 : 0; // codec_config_size_size
+    flags |= itemDataSizeSize == 15 ? 0x400000 : 0;   // item_data_size_size
+    flags |= metadataSizeSize == 10 ? 0x800000 : 0;   // metadata_size_size
 
-    if (image->clli.maxCLL || image->clli.maxPALL || (image->transformFlags != AVIF_TRANSFORM_NONE)) {
-        avifRWStream extendedMetaStream;
-        avifRWStreamStart(&extendedMetaStream, extendedMeta);
-        AVIF_CHECKRES(avifImageWriteExtendedMeta(image, &extendedMetaStream));
-        avifRWStreamFinishWrite(&extendedMetaStream);
-    }
-    AVIF_CHECKRES(avifRWStreamWriteBits(s, extendedMeta->size ? 1 : 0, 1)); // bit(1) has_extended_meta;
-    if (extendedMeta->size) {
-        AVIF_CHECKRES(avifRWStreamWriteVarInt(s, (uint32_t)extendedMeta->size - 1)); // varint extended_meta_size_minus_one;
+    avifBoxMarker meta;
+    AVIF_CHECKRES(avifRWStreamWriteFullBox(s, "meta", AVIF_BOX_SIZE_TBD, /*version=*/1, flags, &meta));
+
+    AVIF_CHECKRES(avifRWStreamWriteBits(s, image->width - 1, dimensionSize));  // unsigned int(dimension_size) width_minus_one;
+    AVIF_CHECKRES(avifRWStreamWriteBits(s, image->height - 1, dimensionSize)); // unsigned int(dimension_size) height_minus_one;
+
+    if (hasIcc) {
+        AVIF_CHECKRES(avifRWStreamWriteBits(s, (uint32_t)image->icc.size - 1, metadataSizeSize)); // unsigned int(metadata_size_size) icc_data_size_minus_one;
+    }
+
+    if (hasExplicitCicp) {
+        AVIF_CHECKRES(avifRWStreamWriteBits(s, image->colorPrimaries, 8));          // bit(8) colour_primaries;
+        AVIF_CHECKRES(avifRWStreamWriteBits(s, image->transferCharacteristics, 8)); // bit(8) transfer_characteristics;
+    }
+    if (chromaSubsampling == 0) {
+        AVIF_CHECKERR(image->matrixCoefficients == AVIF_MATRIX_COEFFICIENTS_UNSPECIFIED, AVIF_RESULT_ENCODE_COLOR_FAILED);
+    } else if (hasExplicitCicp) {
+        AVIF_CHECKRES(avifRWStreamWriteBits(s, image->matrixCoefficients, 8)); // bit(8) matrix_coefficients;
+    }
+
+    AVIF_CHECKRES(avifRWStreamWriteBits(s, codecConfigSize, codecConfigSizeSize)); // unsigned int(codec_config_size_size) main_item_codec_config_size;
+    AVIF_CHECKRES(avifRWStreamWriteBits(s, (uint32_t)colorData->size - 1, itemDataSizeSize)); // unsigned int(item_data_size_size) main_item_data_size;
```

Comment:
Search for `(uint32_t)` casts. Make sure the value is <= `UINT32_MAX` before casting.

Reply:
This should be enforced by `avifEncoderIsMetaBoxV1Compatible()`.